File: 백준/Gold/3190. 뱀/뱀.py
import sys
input = sys.stdin.readline

n = int(input())
k = int(input())

# 맵 만들고 / 사과가 있는 곳은 1로 변경
board = [[0] * n for _ in range(n)]
for i in range(k):
    x, y = map(int, input().split())
    board[x-1][y-1] = 1

# 현재 뱀 꼬리가 어떤 좌표에 있는지 나타냄
visited = [[False] * n for _ in range(n)]

l = int(input())
dict = {}
for i in range(l):
    a, b = map(str, input().split())
    dict[int(a)] = b

dx = [0, 1, 0, -1]
dy = [1, 0, -1, 0]

x, y = 0, 0
cnt = 0
direction = 0

q = [(0,0)]
visited[0][0] = True

while True:
    cnt += 1
    x += dx[direction]
    y += dy[direction]

    # 맵 밖을 벗어나면 종료
    if x < 0 or x >= n or y < 0 or y >= n:
        break

    # 사과를 먹었고, 뱀의 몸이 없는 곳이라면
    if board[x][y] == 1 and visited[x][y] == False:
        # 사과를 제거해주고
        board[x][y] = 0
        # queue에 좌표 입력 / 방문표시
        q.append((x,y))
        visited[x][y] = True

        # 방향 전환
        if cnt in dict:
            if dict[cnt] == 'D':
                direction = (direction + 1) % 4
            elif dict[cnt] == 'L':
                direction = (direction - 1) % 4
    
    # 사과가 없는 곳이고 , 뱀의 몸이 없는 곳이라면
    elif board[x][y] == 0 and visited[x][y] == False:
        # queue에 좌표 입력 / 방문표시
        q.append((x,y))
        visited[x][y] = True
        # 꼬리부분 queue에서 제거 / 방문최신화
        nx, ny = q.pop(0)
        visited[nx][ny] = False

        # 방향 전환
        if cnt in dict:
            if dict[cnt] == 'D':
                direction = (direction + 1) % 4
            elif dict[cnt] == 'L':
                direction = (direction - 1) % 4        
    
    else:
        break

print(cnt)

File: 백준/Gold/3190. 뱀/뱀.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.LinkedList;
import java.util.Queue;
import java.util.StringTokenizer;

public class Main {
    static int n;
    static Apple[] apples;
    static RotateInfo[] rotateInfos;
    static boolean[][] visited;
    // 오른쪽, 아래쪽, 왼쪽, 위쪽
    static int x = 0;
    static int y = 0;
    static int[] dx = {0, 1, 0, -1};
    static int[] dy = {1, 0, -1, 0};
    static int ans = 0;

    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        n = Integer.parseInt(br.readLine());
        int k = Integer.parseInt(br.readLine());

        apples = new Apple[k];
        StringTokenizer st;
        for (int i = 0; i < k; i++) {
            st = new StringTokenizer(br.readLine());
            int a = Integer.parseInt(st.nextToken()) - 1;
            int b = Integer.parseInt(st.nextToken()) - 1;
            apples[i] = new Apple(a, b);
        }

        int l = Integer.parseInt(br.readLine());
        rotateInfos = new RotateInfo[l];
        for (int i = 0; i < l; i++) {
            st = new StringTokenizer(br.readLine());
            int a = Integer.parseInt(st.nextToken());
            String b = st.nextToken();
            rotateInfos[i] = new RotateInfo(a, b);
        }

        visited = new boolean[n][n];
        visited[0][0] = true;

        Queue<int[]> q = new LinkedList<>();
        q.add(new int[]{0, 0});
        Queue<int[]> tail = new LinkedList<>();
        tail.add(new int[]{0, 0});

        while (!q.isEmpty()) {
            int[] arr = q.poll();
            int cnt = arr[0];
            int direction = arr[1];

            int nx = x + dx[direction];
            int ny = y + dy[direction];
            
            if (direction == 3 && nx == -1) {
                ans = cnt + 1;
                break;
            } else if (direction == 1 && nx == n) {
                ans = cnt + 1;
                break;
            } else if (direction == 2 && ny == -1) {
                ans = cnt + 1;
                break;
            } else if (direction == 0 && ny == n) {
                ans = cnt + 1;
                break;
            }

            if (visited[nx][ny]) {
                ans = cnt + 1;
                break;
            }
            boolean isApple = false;
            for (Apple apple : apples) {
                if (apple.x == nx && apple.y == ny && !apple.eaten) {
                    apple.eaten = true;
                    visited[nx][ny] = true;
                    isApple = true;
                    tail.add(new int[]{nx, ny});
                    break;
                }
            }

            // 사과 없을 때
            if (!isApple) {
                visited[nx][ny] = true;
                int[] tailArr = tail.poll();
                int tailX = tailArr[0];
                int tailY = tailArr[1];
                // visited 꼬리 = false;
                visited[tailX][tailY] = false;
                tail.add(new int[]{nx, ny});
            }

            cnt++;

            boolean isRotated = false;
            // 방향 전환
            for (RotateInfo rotateInfo : rotateInfos) {
                if (rotateInfo.time == cnt) {
                    int nowDirection = rotate(direction, rotateInfo.direction);
                    q.add(new int[]{cnt, nowDirection});
                    x = nx;
                    y = ny;
                    isRotated = true;
                    break;
                }
            }

            if (!isRotated) {
                q.add(new int[]{cnt, direction});
                x = nx;
                y = ny;
            }
        }

        System.out.println(ans);
    }

    static class Apple {
        int x;
        int y;
        boolean eaten = false;

        public Apple(int x, int y) {
            this.x = x;
            this.y = y;
        }
    }

    static class RotateInfo {
        int time;
        String direction;

        public RotateInfo(int time, String direction) {
            this.time = time;
            this.direction = direction;
        }
    }

    static int rotate(int num, String direction) {
        if (direction.equals("D")) {
            return (num + 1) % 4;
        } else {
            return (num + 3) % 4;
        }
    }
}